base/expr.jl

```diff
@@ -193,34 +198,89 @@ This is shown in the following example:
         Function Definition
     =#
 end
+
+bigfunction2() do
+    @inline
+    #=
+        Do Body
+    #=
 ```
+
+!!! compat "Julia 1.7"
+    Usage in `do` blocks requires at least Julia 1.7
 """
 macro inline(ex)
     esc(isa(ex, Expr) ? pushmeta!(ex, :inline) : ex)
 end
 
+macro inline()
+    Expr(:meta, :inline)
+end
+
+
 """
     @noinline
 
 Give a hint to the compiler that it should not inline a function.
 
 Small functions are typically inlined automatically.
 By using `@noinline` on small functions, auto-inlining can be
-prevented. This is shown in the following example:
+prevented. `@noinline` can be used at function definitions,
+function calls, and in `do` blocks.
+
+If `@noinline` is used at the beginning of a `do` block of a function
+call, the `do` block won't be inlined but the function itself still
+could be. Similarily, if `@noinline` is specified immediately before a
+function call with a `do` block, that function won't be inlined
+but the `do` block body could possibly be.
+
+This is shown in the following examples:
 
 ```julia
 @noinline function smallfunction(x)
     #=
         Function Definition
     =#
 end
+
+@noinline previouslydefinedfunction(x)
+
+f() do
+    @noinline
+    #=
+        Do Body
+    #=
 ```
 
 !!! note
     If the function is trivial (for example returning a constant) it might get inlined anyway.
+
+!!! compat "Julia 1.7"
+    Callsite usage requires at least Julia 1.7
+
+!!! compat "Julia 1.7"
+    Usage in `do` blocks requires at least Julia 1.7
 """
 macro noinline(ex)
-    esc(isa(ex, Expr) ? pushmeta!(ex, :noinline) : ex)
+    if isa(ex, Expr)
+        if ex.head === :function || is_short_function_def(ex) || ex.head === :->
```

Comment:
```suggestion
        ex = macroexpand(__module__, ex)
        if ex.head === :function || is_short_function_def(ex) || ex.head === :->
```

We need to expand inner macros here so that we still support some patterns like:
```julia
@noinline @aggressive_constprop f(a) = 2sin(a)
```